# Alignment and Counting driver 
This is the main driving script which performs basic QC operations on given reads (namely, trimming and filtering Ribosomal RNA). The driver script (`driver.py`) is designed to be self contained but has following prerequisites:
[Seaborn](https://seaborn.pydata.org/) graphing library
[Matplotlib](https://matplotlib.org/) graphing library (is a prerequisite for Seaborn)
[Pandas](http://pandas.pydata.org/) data analysis library 
R with GGPLOT2
[FASTQCR](https://github.com/kassambara/fastqcr) for plotting 
(Both R and FASTQCR are not required for analysis. The main plots are generated by soon to be uploaded Graphicore)
[Samtools](http://www.htslib.org/) for SAM to BAM conversion
And Python!

The script downloads and uses following packages: 
[HISAT2](https://ccb.jhu.edu/software/hisat2/index.shtml) - Primary alignment package 
[BBDuk](http://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbduk-guide/) - Primary QC package 
[FASTQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) - Secondary QC package
[Subread](http://subread.sourceforge.net/) - For FeatureCount 

    

Driver.py
---------
This script takes input in form of a config file (`config.txt`) which defines following parameters: 
Line 1 : *Threads* (INT)Number of concurrent threads to use
Line 2 : *Read type* (paired or unpaired) *Dummy field*
Line 3 : *Direction* (RF or FR)
This are non-functional
Line 4 : *Front clipping* (INT)
Line 5 : *Reverse clipping* (INT)
Line 6 : *File path of reads* from line 6 onward. 
`Driver.py` is designed for paired end reads, and thus the read path should be in format : `./reads/prefix` and naming convention of paired reads should be :  `prefix_1.fastq.gz` and `prefix_2.fastq.gz`

This is how a sample work directory would look like. `reference` folder should store the reference genome with same format as mentioned below, and should also include the `mrrna.fasta` which contains ribosomal RNA sequences which will be filtered out from raw reads. 

    Workfolder:.
    │   config.txt
    │   driver.py
    │   tree.txt
    │   url.txt
    │       
    ├───reads
    │       Readfiles
    │       
    └───reference
            genome_snp_tran.1.ht2
            genome_snp_tran.2.ht2
            genome_snp_tran.3.ht2
            genome_snp_tran.4.ht2
            genome_snp_tran.5.ht2
            genome_snp_tran.6.ht2
            genome_snp_tran.7.ht2
            genome_snp_tran.8.ht2
            make_grcm38_snp_tran.sh
            mrrna.fasta
    
Alignments can be found in `./alignment` and read count in `./counts` directory. After successful execution of script, the directory structure should look like this:

    Workspace:.
    ├───alignment
    ├───annotation
    ├───bbmap
    │   ├───config
    │   ├───current
    │   │   ├───align2
    │   │   ├───assemble
    │   │   ├───bloom
    │   │   ├───clump
    │   │   ├───cluster
    │   │   ├───dna
    │   │   ├───driver
    │   │   ├───fileIO
    │   │   ├───fun
    │   │   ├───hiseq
    │   │   ├───jgi
    │   │   ├───kmer
    │   │   ├───ml
    │   │   ├───pacbio
    │   │   ├───server
    │   │   ├───shared
    │   │   ├───sketch
    │   │   ├───sort
    │   │   ├───stream
    │   │   ├───structures
    │   │   ├───tax
    │   │   ├───ukmer
    │   │   ├───var
    │   │   └───var2
    │   ├───docs
    │   │   └───guides
    │   ├───jni
    │   ├───pipelines
    │   └───resources
    ├───counts
    ├───FastQC
    │   ├───Configuration
    │   ├───Help
    │   │   ├───1 Introduction
    │   │   ├───2 Basic Operations
    │   │   └───3 Analysis Modules
    │   ├───net
    │   │   └───sourceforge
    │   │       └───iharder
    │   │           └───base64
    │   ├───org
    │   │   └───apache
    │   │       └───commons
    │   │           └───math3
    │   │               ├───analysis
    │   │               │   └───solvers
    │   │               ├───distribution
    │   │               ├───exception
    │   │               │   └───util
    │   │               ├───random
    │   │               ├───special
    │   │               └───util
    │   ├───Templates
    │   │   └───Icons
    │   └───uk
    │       └───ac
    │           └───babraham
    │               └───FastQC
    │                   ├───Analysis
    │                   ├───Dialogs
    │                   ├───FileFilters
    │                   ├───Graphs
    │                   ├───Help
    │                   ├───Modules
    │                   │   └───GCModel
    │                   ├───Report
    │                   ├───Resources
    │                   ├───Results
    │                   ├───Sequence
    │                   │   ├───Contaminant
    │                   │   └───QualityEncoding
    │                   ├───Statistics
    │                   └───Utilities
    ├───hisat
    │   ├───doc
    │   ├───example
    │   │   ├───index
    │   │   ├───reads
    │   │   └───reference
    │   ├───hisatgenotype_modules
    │   ├───hisatgenotype_scripts
    │   └───scripts
    ├───readProcessed
    ├───reads
    ├───reference
    ├───report
    ├───splice
    └───subread
        ├───annotation
        ├───bin
        │   └───utilities
        ├───doc
        └───test
            ├───exactSNP
            │   └───data
            ├───featureCounts
            │   └───data
            ├───subjunc
            │   └───data
            └───subread-align
                └───data




 

